regression/cbmc-incr-smt2/bitvector-bitwise-operators/shift_right.c

```diff
@@ -0,0 +1,36 @@
+// In C, whether a right-shift is arithmetic or logical depends on the
+// original type being shifted. An unsigned value will be shifted to
+// the right in a logical manner (this assigns `0` to the leftmost bit).
+// If the type is signed, right shift will assign the sign bit to the
+// leftmost digit.
+
+int main()
+{
+  int first_pos;
+  __CPROVER_assume(first_pos > 0);
+  int first_neg;
+  __CPROVER_assume(first_neg < 0);
+  unsigned int second;
+
+  int place;
+  __CPROVER_assume(place >= 1);
+
+  int result_signed_pos = first_pos >> place;
+  int result_signed_neg = first_neg >> place;
+  unsigned int result_unsigned = second >> place;
+
+  // This assertion captures the intend of the expected behaviour of
+  // bit-shifting an unsigned int (logical shift)
+  __CPROVER_assert(
+    result_unsigned >= 0,
```

Comment:
Surely this condition would hold for any unsigned value and therefore whether or not this condition holds does not reflect whether our model of `>>` for unsigned ints is correct?

Reply:
Maybe the condition `result_unsigned < second` would be useful and fit into the kind of test you are trying to write. For example in 8 bits, if we take the input `1100 0000` (192) then the correct/logical shift result would be `0110 0000` (96), where as if we take the incorrect arithmetic shift the result would be `1110 0000` (224). Therefore it will be possible to violate `result_unsigned < second` if the wrong kind of shift is used.

Reply:
Given that the old assertion is gone, I consider this to be resolved.